File: hdf-generators/src/delta-report.ts
import type { DeltaResult, LinkRecord } from './delta-types.js';

/**
 * JSON report payload for delta operations.
 * SAF CLI-compatible: { links: LinkRecord[] }
 */
export interface DeltaJsonReport {
  links: LinkRecord[];
}

/**
 * Generate a structured JSON report from a delta result.
 * Matches SAF CLI's delta.json format: { ...diff, links }.
 */
export function generateDeltaJson(result: DeltaResult): DeltaJsonReport {
  return {
    links: result.linkRecords,
  };
}

/**
 * Format a per-link match method description matching SAF CLI's logMatchMethod.
 */
function formatMatchMethod(lr: LinkRecord): string {
  const confidencePct = (lr.confidence * 100).toFixed(0) + '%';
  switch (lr.matchMethod) {
    case 'srgDeterministic':
      return `SRG deterministic (${lr.srg ?? '?'}) [${lr.relationship}]`;
    case 'srgCciTiebreak':
      return `SRG block + CCI tiebreak (Jaccard=${confidencePct}) [${lr.relationship}]`;
    case 'vendorFuzzyTitle':
      return `Vendor fuzzy title (confidence=${confidencePct}) [${lr.relationship}]`;
    case 'exactId':
      return `Exact ID [${lr.relationship}]`;
    case 'cciMatch':
      return `CCI match [${lr.relationship}]`;
    case 'fuzzyTitle':
      return `Fuzzy title (confidence=${confidencePct}) [${lr.relationship}]`;
    case 'none':
      return 'No match';
    default:
      return `${lr.matchMethod} [${lr.relationship}]`;
  }
}

/**
 * Generate a Markdown report from a delta result.
 * Matches SAF CLI's delta.md format: mapping table, control counts,
 * match statistics, and statistics validation.
 */
export function generateDeltaMarkdown(result: DeltaResult): string {
  const lines: string[] = [];
  const stats = result.statistics;

  // Mapping results — SAF CLI format: Old Control -> New Control
  if (result.linkRecords.length > 0) {
    lines.push('Mapping Results ===========================================================================');
    lines.push('\tOld Control -> New Control');
    for (const lr of result.linkRecords) {
      if (lr.relationship !== 'no-match' && lr.oldId) {
        lines.push(`\t   ${lr.oldId} -> ${lr.newId}`);
      }
    }

    const totalMapped = stats.totalMappedControls;
    lines.push(`Total Mapped Controls:  ${totalMapped}`);
    lines.push('');
  }

  // Control counts
  lines.push('Control Counts ===========================');
  lines.push(`Total Controls Available for Delta:  ${stats.oldControlsLength}`);
  lines.push(`     Total Controls Found on XCCDF:  ${stats.newControlsLength}`);
  lines.push('');

  // Match statistics
  lines.push('Match Statistics =========================');
  lines.push(`                    Match Controls:  ${stats.match}`);
  lines.push(`        Possible Mismatch Controls:  ${stats.posMisMatch}`);
  lines.push(`            Related Match Controls:  ${stats.dupMatch}`);
  lines.push(`                 No Match Controls:  ${stats.noMatch}`);
  lines.push('');

  // Statistics validation
  lines.push('Statistics Validation =============================================');
  const totalMapped = stats.totalMappedControls;
  const matchMappedValid = (stats.match + stats.posMisMatch + stats.dupMatch) === totalMapped;
  const processedValid = (totalMapped + stats.noMatch) === stats.newControlsLength;
  lines.push(`Match + Mismatch + Related = Total Mapped Controls:  (${stats.match}+${stats.posMisMatch}+${stats.dupMatch}=${totalMapped}) ${matchMappedValid}`);
  lines.push(`  Total Processed = Total XCCDF Controls:  (${totalMapped}+${stats.noMatch}=${totalMapped + stats.noMatch}) ${processedValid}`);
  lines.push('');

  // Per-control match method details
  if (result.linkRecords.length > 0) {
    lines.push('Match Details =============================================================');
    for (const lr of result.linkRecords) {
      if (lr.oldId) {
        lines.push(`  ${lr.oldId} --> ${lr.newId}`);
        lines.push(`       Match method:  ${formatMatchMethod(lr)}`);
      } else {
        lines.push(`  (none) --> ${lr.newId}  [no match]`);
      }
    }
    lines.push('');
  }

  return lines.join('\n');
}

File: hdf-generators/go/delta_report.go
package generators

import (
	"encoding/json"
	"fmt"
	"strings"
)

// DeltaJsonReport is the JSON report payload (SAF CLI-compatible).
type DeltaJsonReport struct {
	Links []LinkRecord `json:"links"`
}

// GenerateDeltaJSON generates a structured JSON report from a delta result.
func GenerateDeltaJSON(result DeltaResult) ([]byte, error) {
	report := DeltaJsonReport{
		Links: result.LinkRecords,
	}
	return json.MarshalIndent(report, "", "  ")
}

// formatMatchMethod returns a SAF CLI-compatible match method description.
func formatMatchMethod(lr LinkRecord) string {
	confidencePct := fmt.Sprintf("%d%%", int(lr.Confidence*100))
	switch lr.MatchMethod {
	case "srgDeterministic":
		srg := lr.SRG
		if srg == "" {
			srg = "?"
		}
		return fmt.Sprintf("SRG deterministic (%s) [%s]", srg, lr.Relationship)
	case "srgCciTiebreak":
		return fmt.Sprintf("SRG block + CCI tiebreak (Jaccard=%s) [%s]", confidencePct, lr.Relationship)
	case "vendorFuzzyTitle":
		return fmt.Sprintf("Vendor fuzzy title (confidence=%s) [%s]", confidencePct, lr.Relationship)
	case "exactId":
		return fmt.Sprintf("Exact ID [%s]", lr.Relationship)
	case "cciMatch":
		return fmt.Sprintf("CCI match [%s]", lr.Relationship)
	case "fuzzyTitle":
		return fmt.Sprintf("Fuzzy title (confidence=%s) [%s]", confidencePct, lr.Relationship)
	case "none":
		return "No match"
	default:
		return fmt.Sprintf("%s [%s]", lr.MatchMethod, lr.Relationship)
	}
}

// GenerateDeltaMarkdown generates a Markdown report matching SAF CLI's format.
func GenerateDeltaMarkdown(result DeltaResult) string {
	var b strings.Builder
	stats := result.Statistics

	// Mapping results
	if len(result.LinkRecords) > 0 {
		b.WriteString("Mapping Results ===========================================================================\n")
		b.WriteString("\tOld Control -> New Control\n")
		for _, lr := range result.LinkRecords {
			if lr.Relationship != "no-match" && lr.OldID != "" {
				fmt.Fprintf(&b, "\t   %s -> %s\n", lr.OldID, lr.NewID)
			}
		}
		fmt.Fprintf(&b, "Total Mapped Controls:  %d\n\n", stats.TotalMappedControls)
	}

	// Control counts
	b.WriteString("Control Counts ===========================\n")
	fmt.Fprintf(&b, "Total Controls Available for Delta:  %d\n", stats.OldControlsLength)
	fmt.Fprintf(&b, "     Total Controls Found on XCCDF:  %d\n\n", stats.NewControlsLength)

	// Match statistics
	b.WriteString("Match Statistics =========================\n")
	fmt.Fprintf(&b, "                    Match Controls:  %d\n", stats.Match)
	fmt.Fprintf(&b, "        Possible Mismatch Controls:  %d\n", stats.PosMisMatch)
	fmt.Fprintf(&b, "            Related Match Controls:  %d\n", stats.DupMatch)
	fmt.Fprintf(&b, "                 No Match Controls:  %d\n\n", stats.NoMatch)

	// Statistics validation
	b.WriteString("Statistics Validation =============================================\n")
	totalMapped := stats.TotalMappedControls
	matchMappedValid := (stats.Match + stats.PosMisMatch + stats.DupMatch) == totalMapped
	processedValid := (totalMapped + stats.NoMatch) == stats.NewControlsLength
	fmt.Fprintf(&b, "Match + Mismatch + Related = Total Mapped Controls:  (%d+%d+%d=%d) %t\n",
		stats.Match, stats.PosMisMatch, stats.DupMatch, totalMapped, matchMappedValid)
	fmt.Fprintf(&b, "  Total Processed = Total XCCDF Controls:  (%d+%d=%d) %t\n\n",
		totalMapped, stats.NoMatch, totalMapped+stats.NoMatch, processedValid)

	// Per-control match details
	if len(result.LinkRecords) > 0 {
		b.WriteString("Match Details =============================================================\n")
		for _, lr := range result.LinkRecords {
			if lr.OldID != "" {
				fmt.Fprintf(&b, "  %s --> %s\n", lr.OldID, lr.NewID)
				fmt.Fprintf(&b, "       Match method:  %s\n", formatMatchMethod(lr))
			} else {
				fmt.Fprintf(&b, "  (none) --> %s  [no match]\n", lr.NewID)
			}
		}
		b.WriteString("\n")
	}

	return b.String()
}
